File: design_patterns/factory/queue.rb
require_relative './list.rb'
require_relative './array.rb'

class Queue
  def self.create(type)
    case type
    when 'list'
      return List.new()
    when 'array'
      return Array.new()
    else
      raise 'Unvalid queue type #{type}, Types are : list, array.'
    end
  end
end

File: design_pattern/queue/Queue.php
<?php

/**
 * Queue
 */
class Queue {

  public static function create ($type) {
    switch ($type) {
      case 'list':
        return new List();
      case 'array':
        return new Array();
      default:
        throw new Exception("Undefined Queue type", 1);
    }
  }

}